### PR TITLE
[RFR][Feature][OSF-5939]Sort files by Dates in project(overview) page and project files page

### PR DESCRIPTION
**Pull request by TomBaxter**
## Purpose:
[OSF-5939](https://openscience.atlassian.net/browse/OSF-5939)
In project overview page and project file view page...
- Add last modified date to file info.
- Allow sorting by modified date

## Bonus
GitHub files now show their file size in project files view
Dataverse Raven logging should now work

## Changes:
Update website/addons/dataverse/static/dataverseFangornConfig.js
- Require raven-js to resolve unknown name
- Add columns size, downloads, and modified to allow for sorting by last modified date

Update website/addons/github/static/githubFangornConfig.js
- Minor clean up
- Add columns size, downloads, and modified to allow for sorting by last modified date

Update  website/static/js/fangorn.js 
- Require moment-js for handling date/time
- Add  function _fangornModifiedColumn to normalize modified date using moment-js
- Update function _fangornResolveRows to fill in size, download, and modified for all files
- Allow for sorting by last modified date

Update website/static/js/pages/project-dashboard-page.js
- Update $(document).ready(function () to show sortable modified column

## Side effects
None

## Caveats
WB does not provide file last modified dates for the following providers...
Dataverse
FigShare
GitHub

WB does not provide file sizes for the Dataverse provider

WB only provides download counts for the OSFStorage provider

[#OSF-5939]

### CODE REVIEW

**Comment by mfraezz** [comment]
Looks good on the project dashboard:

screen shot 2016-06-10 at 16 50 17

However, there's a 🐛 with the files page. The requests and responses seem to be going out correctly, but the UI is broken:

screen shot 2016-06-10 at 16 52 16

File: website/addons/dataverse/static/dataverseFangornConfig.js [hunk 1]
```diff
@@ -254,25 +255,36 @@ function _fangornDataverseTitle(item, col) {
...
     columns.push({
         data : 'name',
         folderIcons : true,
         filter : true,
-        css: selectClass,
         custom: _fangornDataverseTitle
     });
-
-    if (tb.options.placement === 'project-files') {
+    if(tb.options.placement === 'project-files') {
```
Comment (mfraezz) [inline website/addons/dataverse/static/dataverseFangornConfig.js]: This line and the other conditional below should both have a space between `if` and `(`, as per [cosdev](http://cosdev.readthedocs.io/en/latest/style_guides/javascript.html)

**Comment by mfraezz** [comment]
Pass complete, please also resolve the merge conflicts. Back to you @TomBaxter 🐠

**Comment by TomBaxter** [comment]
File view fixed for non-existant item.data.extra.downloads effecting folders in OSFStorage.
Removed console statements.
Fixed no space between "if" and "(".
Fixed merge conflict.
Ready for further review.

File: website/addons/github/static/githubFangornConfig.js [hunk 2]
```diff
@@ -340,11 +337,27 @@ function _fangornColumns (item) {
...
         columns.push(
         {
+            data  : 'size',
+            sortInclude : false,
+            filter : false,
+            custom : function() {return item.data.size ? $osf.humanFileSize(item.data.size, true) : '';}
+        });
+    }
+
+    if(tb.options.placement === 'project-files') {
```
Comment (mfraezz) [inline website/addons/github/static/githubFangornConfig.js]: This looks like the same conditional as above; could you combine the `'downloads'` and `'size'` data to the same `.push` call?

**Comment by mfraezz** [comment]
Minor: I'm getting a `TypeError: item is undefined` from [here](https://github.com/CenterForOpenScience/osf.io/blob/9bb285a705f106dd25197dadb761c2f1790fa9ca/website/static/js/fangorn.js#L2374), but it doesn't seem to be causing any behavior issues. This happens when expanding/collapsing folders (including addon root folders), but only when no rows are 'highlighted'. Confirmed that previously documented :bug: was fixed.

screen shot 2016-06-14 at 14 11 46

**Comment by TomBaxter** [comment]
The TypeError was introduced with [PR#5835](https://github.com/CenterForOpenScience/osf.io/pull/5835). Looking into a fix now.

File: website/addons/github/static/githubFangornConfig.js (hunk 2 above)
Comment (TomBaxter) [inline website/addons/github/static/githubFangornConfig.js]: Done.

**Comment by erinspace** [comment]
@TomBaxter #5835 should have *fixed* that typeError on expanding and contracting

**Comment by TomBaxter** [comment]
@erinspace OK, I'll double check.

**Comment by TomBaxter** [comment]
@erinspace You right, sorry about that. Apparently I have dyslexia, but with red and green.

File: website/addons/dataverse/static/dataverseFangornConfig.js [hunk 3]
```diff
@@ -254,25 +255,36 @@ function _fangornDataverseTitle(item, col) {
...
-            }
-        );
+        {
+            data  : 'size',
+            sortInclude : false,
+            filter : false,
+            custom : function() {return m('');}
+        });
     }
+    if (tb.options.placement === 'project-files') {
```
Comment (mfraezz) [inline website/addons/dataverse/static/dataverseFangornConfig.js]: Ditto [this comment](https://github.com/CenterForOpenScience/osf.io/pull/5817/files#r67015172)

>Minor: this looks like the same conditional as above; could you combine the 'downloads' and 'size' data to the same .push call?

Perhaps also consider DRYing this code up, but that's not a blocker.

Comment (TomBaxter) [inline website/addons/dataverse/static/dataverseFangornConfig.js]: Done and pushed.

File: website/static/js/fangorn.js [hunk 4]
```diff
@@ -10,6 +10,7 @@ var m = require('mithril');
 var URI = require('URIjs');
 var Raven = require('raven-js');
 var Treebeard = require('treebeard');
+var Moment = require('moment');
```
Comment (sloria) [inline website/static/js/fangorn.js]: Minor: By convention, we do not capitalize `moment`.

File: website/static/js/fangorn.js [hunk 5]
```diff
@@ -1270,6 +1271,30 @@ function _fangornTitleColumn(item, col) {
...
+    var tb = this;
+    if (item.data.isAddonRoot && item.connected === false) { // as opposed to undefined, avoids unnecessary setting of this value
+        return _connectCheckTemplate.call(this, item);
+    }
+    var assert = require('assert');
+    if (item.kind === 'file' && item.data.permissions.view && item.data.modified) {
+        return m(
+            'span',
+            // "new Date" required for non-ISO date formats
+            new Moment(new Date(item.data.modified)).format('YYYY-MM-DD hh:mm A')
```
Comment (sloria) [inline website/static/js/fangorn.js]: From the docs, it doesn't appear correct to do `new Moment`. `moment(new Date(...)).format(...)` should work. Can you verify?

File: website/static/js/fangorn.js [hunk 6]
```diff
@@ -1270,6 +1271,30 @@ function _fangornTitleColumn(item, col) {
...
+ * @this Treebeard.controller
+ * @returns {Array} Returns an array of mithril template objects using m()
+ * @private
+ */
+function _fangornModifiedColumn(item, col) {
+    var tb = this;
+    if (item.data.isAddonRoot && item.connected === false) { // as opposed to undefined, avoids unnecessary setting of this value
+        return _connectCheckTemplate.call(this, item);
+    }
+    var assert = require('assert');
```
Comment (sloria) [inline website/static/js/fangorn.js]: This should be removed.

File: website/static/js/fangorn.js [hunk 7]
```diff
@@ -1340,29 +1365,33 @@ function _fangornResolveRows(item) {
...
+        sortInclude : false,
+        filter : false,
+        custom : function() {return item.data.size ? $osf.humanFileSize(item.data.size, true) : '';}
+    });
+    if (item.data.provider === 'osfstorage') {
+        default_columns.push({
+            data : 'downloads',
+            sortInclude : false,
+            filter : false,
+            custom: function() { return item.data.extra && item.data.extra.downloads ? item.data.extra.downloads.toString() : ''; }
```
Comment (sloria) [inline website/static/js/fangorn.js]: The preferred way to get deeply nested attributes is to use `lodash.get`.

```js
... return lodashGet(item, 'data.extra.downloads') ? ...
```

**Comment by sloria** [comment]
Pass finished.

File: website/static/js/fangorn.js (hunk 5 above)
Comment (TomBaxter) [inline website/static/js/fangorn.js]: That's how I did it originally but I got a jshint error from Travis without the "new"

File: website/static/js/fangorn.js (hunk 4 above)
Comment (TomBaxter) [inline website/static/js/fangorn.js]: Done.

File: website/static/js/fangorn.js (hunk 6 above)
Comment (TomBaxter) [inline website/static/js/fangorn.js]: Done.

File: website/static/js/fangorn.js (hunk 7 above)
Comment (TomBaxter) [inline website/static/js/fangorn.js]: Done, and nifty.

**Comment by sloria** [comment]
Looks good. 🐙

**Pull request merged**